pkg/orchestrator/housekeeping.go

```diff
@@ -0,0 +1,219 @@
+	var activeExecutions []*models.Execution
+	activeJobs, err := h.jobStore.GetInProgressJobs(h.ctx, "")
+	if err != nil {
+		log.Ctx(h.ctx).Err(err).Msg("failed to get active jobs")
+		return activeExecutions
+	}
+
+	for i := range activeJobs {
+		job := &activeJobs[i]
+
+		// we only have housekeeping tasks for batch and ops jobs
+		if job.IsLongRunning() {
+			continue
+		}
+		executions, err := h.jobStore.GetExecutions(h.ctx, jobstore.GetExecutionsOptions{
+			JobID: job.ID,
+		})
+		if err != nil {
+			// log error and avoid having a single job failure affect the housekeeping of other jobs
+			log.Ctx(h.ctx).Err(err).Msgf("failed to get executions for job %s", job.ID)
+			continue
+		}
+		// filter terminal executions, and enrich executions with job information
+		for j := range executions {
+			if executions[j].IsTerminalState() {
+				continue
+			}
+			executions[j].Job = job
+			activeExecutions = append(activeExecutions, &executions[j])
+		}
+	}
+	return activeExecutions
+}
+
+// timeoutExecutions checks for executions that have been in progress beyond the timeout period
+// and enqueue an evaluation for them. It is the responsibility of the scheduler to fail the executions
+func (h *Housekeeping) timeoutExecutions(activeExecutions []*models.Execution) {
+	alreadyEvaluatedJobs := make(map[string]struct{})
+	for _, execution := range activeExecutions {
+		// skip if the job has already been evaluated by another active execution
+		if _, ok := alreadyEvaluatedJobs[execution.JobID]; ok {
+			continue
+		}
+
+		timeoutWithBuffer := execution.Job.Task().Timeouts.GetExecutionTimeout() + h.timeoutBuffer
+		if execution.HasExecutionExpired(timeoutWithBuffer) {
+			// acquire semaphore to limit the number of concurrent housekeeping tasks
+			h.workersSem <- struct{}{}
+			h.waitGroup.Add(1)
+			alreadyEvaluatedJobs[execution.JobID] = struct{}{}
+
+			go func(execution *models.Execution) {
+				defer h.waitGroup.Done()
+				defer func() { <-h.workersSem }() // release semaphore
+				h.handleTimeoutExecutions(execution)
+			}(execution)
+		}
+	}
+}
+
+// handleTimeoutExecutions handles the timeout of an execution
+// TODO: atomic creation and enqueue of evaluations #3972
+func (h *Housekeeping) handleTimeoutExecutions(execution *models.Execution) {
+	// enqueue evaluation to trigger the scheduler to communicate the cancellation to the compute
+	// node, and schedule a new execution if applicable
+	eval := models.NewEvaluation().
+		WithJobID(execution.JobID).
+		WithTriggeredBy(models.EvalTriggerExecTimeout).
+		WithType(execution.Job.Type).
+		WithComment(fmt.Sprintf("execution %s timed out", execution.ID)).
+		Normalize()
+
+	err := h.jobStore.CreateEvaluation(h.ctx, *eval)
+	if err != nil {
+		log.Ctx(h.ctx).Err(err).Msgf("failed to create evaluation %+v", eval)
+		return
+	}
+
+	err = h.evaluationBroker.Enqueue(eval)
+	if err != nil {
+		log.Ctx(h.ctx).Err(err).Msgf("failed to enqueue evaluation %+v", eval)
+		return
+	}
```

Comment:
I would prefer we push these error up the stack to the root caller `runHousekeepingTasks` and deal with them consistently in a single location.

Reply:
done in https://github.com/bacalhau-project/bacalhau/pull/3974/commits/f380e99ba72b60e64aef96166efff56967bc73b1